libraries/AP_RangeFinder/AP_RangeFinder_Benewake_CAN.cpp

```diff
@@ -0,0 +1,74 @@
+#include <AP_HAL/AP_HAL.h>
+#include "AP_RangeFinder_Benewake_CAN.h"
+
+#if HAL_MAX_CAN_PROTOCOL_DRIVERS
+
+const AP_Param::GroupInfo AP_RangeFinder_Benewake_CAN::var_info[] = {
+
+    // @Param: RECV_ID
+    // @DisplayName: CAN receive ID
+    // @Description: The receive ID of the CAN frames. A value of zero means all IDs are accepted.
+    // @Range: 0 65535
+    // @User: Advanced
+    AP_GROUPINFO("RECV_ID", 10, AP_RangeFinder_Benewake_CAN, receive_id, 0),
+
+    // @Param: SNR_MIN
+    // @DisplayName: Minimum signal strength
+    // @Description: Minimum signal strength (SNR) to accept distance
+    // @Range: 0 65535
+    // @User: Advanced
+    AP_GROUPINFO("SNR_MIN", 11, AP_RangeFinder_Benewake_CAN, snr_min, 0),
+
+    AP_GROUPEND
+};
+
+
+/*
+  constructor
+ */
+AP_RangeFinder_Benewake_CAN::AP_RangeFinder_Benewake_CAN(RangeFinder::RangeFinder_State &_state, AP_RangeFinder_Params &_params) :
+    CANSensor("Benewake"),
+    AP_RangeFinder_Backend(_state, _params)
+{
+    register_driver(AP_CANManager::Driver_Type_Benewake);
+    AP_Param::setup_object_defaults(this, var_info);
+    state.var_info = var_info;
+}
+
+// update state
+void AP_RangeFinder_Benewake_CAN::update(void)
+{
+    WITH_SEMAPHORE(_sem);
+    const uint32_t now = AP_HAL::millis();
+    if (_distance_count == 0 && now - state.last_reading_ms > 500) {
+        // no new data.
+        set_status(RangeFinder::Status::NoData);
+    } else if (_distance_count != 0) {
+        state.distance_m = _distance_sum / _distance_count;
+        state.last_reading_ms = AP_HAL::millis();
+        _distance_sum = 0;
+        _distance_count = 0;
+        update_status();
+    }
+}
+
+// handler for incoming frames. These come in at 100Hz
+void AP_RangeFinder_Benewake_CAN::handle_frame(AP_HAL::CANFrame &frame)
+{
+    WITH_SEMAPHORE(_sem);
+    const uint16_t id = frame.id & AP_HAL::CANFrame::MaskStdID;
+    if (receive_id != 0 && id != uint16_t(receive_id.get())) {
```

Comment:
Should we lock in on the first receive_id we receive, rather than accepting any and all?  At least people attempting to put multiple on the same bus will get a consistent reading (if from a random sensor)

Reply:
done